### DISCUSSION TITLE
Prehook & pass data to `handleDataRequest` to auto set-cookie for refreshed session.

### DISCUSSION THREAD

**Discussion by WesleyKapow** [remix-run/remix#3377] [Proposals]
This is related to https://github.com/remix-run/remix/discussions/2166 and https://github.com/sergiodxa/remix-auth/discussions/147.

The basic idea is that we need to return the `Set-Cookie` header with a new session whenever an auth token is refreshed (via a refresh token). The current best solution is to either do this outside of remix (express, if you are using it) OR have all of your loaders & actions do something like:

```ts
export const action: ActionFunction = async ({ request }) => {
  const { headers, accessToken } = await authenticateSession(request);
  const data = doSomeRequests(accessToken)
  return json(data, { headers }); // Return headers that may contain 'Set-Cookie'
};
```

NOTE: Loaders can avoid dealing with `headers` themselves as we can do a `throw redirect(request.url, {headers}` within `authenticateSession`.

The two sad parts here are:
1. Every loader/action has to manually authenticate the session.
2. Actions need to remember to return the possibly updated headers.

Number 1 above can be handled via some type of prehook that I believe is being discussed (I can't find any GH discussion on it).

Number 2 above could be resolved if there was someway for `authenticateSession` to setup something that `handleDataRequest` can read, allowing `handleDataRequest` to always return the possibly updated `headers`.

**Comment by WesleyKapow** [comment]
Oh and if we did have a prehook that could validate the session and possibly refresh the token for us. It'd be good if the prehook can mutate the `request` that the loaders/actions then see, so that they operate using the new session and not the stale one.

**Comment by kiliman** [comment]
There's a Draft PR submitted by @jacob-ebey adding `beforeRequest` and `beforeResponse` handlers

I think it's on hold at the moment. 

https://github.com/remix-run/remix/pull/364
